File: browser/js/core/script/lower.js
// Port of core/script/lower.cpp — statements to the flat op stream.
import { argsFilter, argsFrame, argsLayout, argsSave, argsShape } from './args.js';
import { argsCrop } from './crop.js';
import { hasErrors, makeDiag, tokenOfStmt } from './diagnostics.js';
import { argsUse } from './lineStyle.js';
import { expandStencilUse, reportUnusedTemplates, templateIndex } from './templates.js';
import {
  MAX_OPS, MAX_SOURCE_CHARS, MAX_TEMPLATE_EXPANSIONS, SOURCE_KINDS, classifySource,
  defaultLineStyle, isEditDirective, isStencilUse,
} from './types.js';
import { EditLedger, applyHistoryStmt } from './undo.js';
import { joinWords } from './values.js';

// The edit directives: the op each lowers to, and the grammar that fills it.
const EDIT_OPS = {
  crop: { kind: 'crop', read: (st, state, op, diags) => argsCrop(st, state, op, diags) },
  filter: { kind: 'filter', read: (st, state, op, diags) => argsFilter(st, op, diags) },
  layout: { kind: 'layout', read: (st, state, op, diags) => argsLayout(st, op, diags) },
  line: { kind: 'line', read: (st, state, op, diags) => argsShape(st, state, false, op, diags) },
  rect: { kind: 'rect', read: (st, state, op, diags) => argsShape(st, state, true, op, diags) },
};

// What `@undo @line …` matches on: the directive plus its argument words.
const normalizedText = (s) => {
  let out = s.directive;
  for (const t of s.args) {
    if (t.kind === 'comment') continue;
    out += ` ${t.text.toLowerCase()}`;
  }
  return out;
};

const blankOp = (kind, st, block) => ({
  kind,
  block,
  line: st.line,
  col: st.col,
  len: st.len,
  editIndex: 0,
  strs: [],
  toks: [],
  nums: [],
});

export const lowerScript = (parsed) => {
  const diagnostics = parsed.diagnostics.slice();
  const blocks = [];
  const ops = [];
  const emptyBlocks = [];
  const byName = templateIndex(parsed.templates);
  let capped = false; // a replay that would pass MAX_OPS stops the whole script
  const budget = { used: 0 }; // the whole script fan-out, blocks included

  for (const raw of parsed.blocks) {
    const blockIndex = blocks.length;
    const block = {
      source: '',
      kind: 'project',
      frame: 0,
      opStart: ops.length,
      opCount: 0,
      line: raw.implicit ? 1 : raw.header.line,
    };

    if (!raw.implicit) {
      block.source = joinWords(raw.header.args);
      block.kind = classifySource(block.source);
      if (block.source.length > MAX_SOURCE_CHARS) {
        diagnostics.push(makeDiag('error', 'E_LIMIT_SOURCE', tokenOfStmt(raw.header),
          'the @source spec is too long'));
        block.source = block.source.slice(0, MAX_SOURCE_CHARS);
      }
      const open = blankOp('open', raw.header, blockIndex);
      open.strs = [block.source];
      open.nums = [SOURCE_KINDS.indexOf(block.kind)];
      ops.push(open);
    }

    // Expand templates first, so edit numbering counts what a template contributed.
    const body = [];
    for (const st of raw.body) {
      if (st.directive !== 'use' || !isStencilUse(st)) { body.push(st); continue; }
      expandStencilUse(st, parsed.templates, byName, 1, budget, body, diagnostics);
      if (budget.used > MAX_TEMPLATE_EXPANSIONS) { capped = true; break; }
    }
    if (capped) break; // the capped block is not recorded, so nothing of it dumps

    const state = { unit: 'px', style: defaultLineStyle() };
    const ledger = new EditLedger();
    let sawSave = false;
    let sawEdit = false;
    const frames = [];

    const rewindAndReplay = (line, col) => {
      if (ledger.reconcile(ops, blockIndex, line, col)) return true;
      diagnostics.push(makeDiag('error', 'E_LIMIT_OPS',
        raw.implicit ? null : tokenOfStmt(raw.header), 'the script has too many ops'));
      capped = true;
      return false;
    };

    for (const st of body) {
      const d = st.directive;

      if (d === 'use') {
        argsUse(st, state, diagnostics);
        continue;
      }

      if (d === 'frame') {
        if (raw.implicit) {
          diagnostics.push(makeDiag('error', 'E_FRAME_OUTSIDE_SOURCE', tokenOfStmt(st),
            '@frame needs a @source block naming a video'));
          continue;
        }
        const op = blankOp('frame', st, blockIndex);
        if (!argsFrame(st, op, diagnostics)) continue;
        const idx = op.nums[0];
        if (frames.includes(idx)) {
          diagnostics.push(makeDiag('error', 'E_DUPLICATE_FRAME', tokenOfStmt(st),
            `frame ${idx} is already used in this block`));
          continue;
        }
        frames.push(idx);
        if (!rewindAndReplay(st.line, st.col)) break;
        ledger.reset();
        ops.push(op);
        continue;
      }

      if (d === 'undo' || d === 'redo') {
        applyHistoryStmt(st, d === 'redo', ledger, diagnostics);
        continue;
      }

      if (d === 'save') {
        const op = blankOp('save', st, blockIndex);
        argsSave(st, op);
        if (!rewindAndReplay(st.line, st.col)) break;
        ops.push(op);
        sawSave = true;
        continue;
      }

      if (!isEditDirective(d)) {
        diagnostics.push(makeDiag('error', 'E_BAD_TOKEN', tokenOfStmt(st),
          `'@${d}' cannot be used here`));
        continue;
      }

      const { kind, read } = EDIT_OPS[d];
      const op = blankOp(kind, st, blockIndex);
      if (!read(st, state, op, diagnostics)) continue;
      if (ops.length >= MAX_OPS) {
        diagnostics.push(makeDiag('error', 'E_LIMIT_OPS', tokenOfStmt(st),
          'the script has too many ops'));
        break;
      }
      // Numbered before the ledger copies it, so a replayed edit dumps as the same edit.
      op.editIndex = ledger.editCount + 1;
      ledger.addEdit(op, normalizedText(st));
      ops.push(op);
      sawEdit = true;
    }

    if (!capped) rewindAndReplay(block.line, 1);
    if (capped) break; // the capped block is not recorded, so nothing of it dumps

    if (!raw.implicit && !sawEdit && !sawSave) emptyBlocks.push(raw.header);

    block.opCount = ops.length - block.opStart;
    blocks.push(block);
  }

  reportUnusedTemplates(parsed.templates, diagnostics);
  // A script with an error never runs, so "does nothing" would only add noise.
  if (!hasErrors(diagnostics)) {
    for (const header of emptyBlocks) {
      diagnostics.push(makeDiag('warning', 'W_EMPTY_BLOCK', tokenOfStmt(header),
        'this @source block does nothing'));
    }
  }
  return { blocks, ops, diagnostics };
};

File: core/script/lower.cpp
#include "lower.hpp"

#include "args.hpp"
#include "diagnostics.hpp"
#include "templates.hpp"
#include "undo.hpp"
#include "values.hpp"
#include "text.hpp"

#include <utility>

namespace stencil::core::script {

  namespace {

    // What `@undo @line …` matches on: the directive plus its argument words.
    std::string normalizedText(const Stmt& s) {
      std::string out = s.directive;
      for (const Token& t : s.args) {
        if (t.kind == TokenKind::COMMENT) continue;
        out.push_back(' ');
        out += toLowerAscii(t.text);
      }
      return out;
    }

    Op blankOp(OpKind kind, const Stmt& st, int block) {
      Op op;
      op.kind = kind;
      op.block = block;
      op.line = st.line;
      op.col = st.col;
      op.len = st.len;
      return op;
    }

    bool isStencilUse(const Stmt& st) {
      return st.kind == Directive::USE && !st.args.empty() &&
             toLowerAscii(unquoteWord(st.args[0].text)) == "stencil";
    }

  }  // namespace

  LowerResult lowerScript(ParseResult& parsed) {
    LowerResult out;
    out.diagnostics = std::move(parsed.diagnostics);
    std::vector<Token> emptyBlocks;
    bool capped = false;  // a replay that would pass MAX_OPS stops the whole script
    int expansions = 0;   // the whole script's template fan-out, blocks included

    for (RawBlock& raw : parsed.blocks) {
      const int blockIndex = static_cast<int>(out.blocks.size());

      Block block;
      block.line = raw.implicit ? 1 : raw.header.line;
      block.opStart = static_cast<int>(out.ops.size());
      if (!raw.implicit) {
        block.source = joinWords(raw.header.args);
        block.kind = classifySource(block.source);
        if (static_cast<int>(block.source.size()) > MAX_SOURCE_CHARS) {
          out.diagnostics.push_back(makeDiag(Severity::ERROR, "E_LIMIT_SOURCE", raw.header,
                                             "the @source spec is too long"));
          block.source.resize(MAX_SOURCE_CHARS);
        }
        Op open = blankOp(OpKind::OPEN, raw.header, blockIndex);
        open.strs.assign(1, block.source);
        open.nums.assign(1, static_cast<double>(block.kind));
        out.ops.push_back(std::move(open));
      }

      // Expand templates first, so edit numbering counts what a template contributed.
      std::vector<Stmt> body;
      for (Stmt& st : raw.body) {
        if (!isStencilUse(st)) { body.push_back(std::move(st)); continue; }
        expandStencilUse(st, parsed.templates, 1, expansions, body, out.diagnostics);
        if (expansions > MAX_TEMPLATE_EXPANSIONS) { capped = true; break; }
      }
      if (capped) break;  // the capped block is not recorded, so nothing of it dumps

      EvalState state;
      EditLedger ledger;
      bool sawSave = false, sawEdit = false;
      std::vector<int> frames;

      auto rewindAndReplay = [&](int line, int col) {
        if (ledger.reconcile(out.ops, blockIndex, line, col)) return true;
        out.diagnostics.push_back(makeDiag(Severity::ERROR, "E_LIMIT_OPS", raw.header,
                                           "the script has too many ops"));
        capped = true;
        return false;
      };

      for (Stmt& st : body) {
        if (st.kind == Directive::USE) {
          bool sawStencil = false;
          argsUse(st, state, sawStencil, out.diagnostics);
          continue;
        }

        if (st.kind == Directive::FRAME) {
          if (raw.implicit) {
            out.diagnostics.push_back(makeDiag(Severity::ERROR, "E_FRAME_OUTSIDE_SOURCE", st,
                                               "@frame needs a @source block naming a video"));
            continue;
          }
          Op op = blankOp(OpKind::FRAME, st, blockIndex);
          if (!argsFrame(st, op, out.diagnostics)) continue;
          const int idx = static_cast<int>(op.nums[0]);
          bool isDuplicate = false;
          for (int f : frames)
            if (f == idx) { isDuplicate = true; break; }
          if (isDuplicate) {
            out.diagnostics.push_back(makeDiag(Severity::ERROR, "E_DUPLICATE_FRAME", st,
                                               "frame " + std::to_string(idx) +
                                                   " is already used in this block"));
            continue;
          }
          frames.push_back(idx);
          if (!rewindAndReplay(st.line, st.col)) break;
          ledger.reset();
          out.ops.push_back(std::move(op));
          continue;
        }

        if (st.kind == Directive::UNDO || st.kind == Directive::REDO) {
          applyHistoryStmt(st, st.kind == Directive::REDO, ledger, out.diagnostics);
          continue;
        }

        if (st.kind == Directive::SAVE) {
          Op op = blankOp(OpKind::SAVE, st, blockIndex);
          if (!argsSave(st, op, out.diagnostics)) continue;
          if (!rewindAndReplay(st.line, st.col)) break;
          out.ops.push_back(std::move(op));
          sawSave = true;
          continue;
        }

        if (!isEditDirective(st.kind)) {
          out.diagnostics.push_back(makeDiag(Severity::ERROR, "E_BAD_TOKEN", st,
                                             "'@" + st.directive + "' cannot be used here"));
          continue;
        }

        Op op;
        bool ok = false;
        switch (st.kind) {
          case Directive::CROP:
            op = blankOp(OpKind::CROP, st, blockIndex);
            ok = argsCrop(st, state, op, out.diagnostics);
            break;
          case Directive::FILTER:
            op = blankOp(OpKind::FILTER, st, blockIndex);
            ok = argsFilter(st, op, out.diagnostics);
            break;
          case Directive::LAYOUT:
            op = blankOp(OpKind::LAYOUT, st, blockIndex);
            ok = argsLayout(st, op, out.diagnostics);
            break;
          default: {  // @line, and @rect as a line with locked corners
            const bool locked = st.kind == Directive::RECT;
            op = blankOp(locked ? OpKind::RECT : OpKind::LINE, st, blockIndex);
            ok = argsShape(st, state, locked, op, out.diagnostics);
            break;
          }
        }
        if (!ok) continue;
        if (static_cast<int>(out.ops.size()) >= MAX_OPS) {
          out.diagnostics.push_back(
              makeDiag(Severity::ERROR, "E_LIMIT_OPS", st, "the script has too many ops"));
          break;
        }
        // Numbered before the ledger copies it, so a replayed edit dumps as the same edit.
        op.editIndex = ledger.editCount() + 1;
        ledger.addEdit(op, normalizedText(st));
        out.ops.push_back(std::move(op));
        sawEdit = true;
      }

      if (!capped) rewindAndReplay(block.line, 1);
      if (capped) break;  // the capped block is not recorded, so nothing of it dumps

      if (!raw.implicit && !sawEdit && !sawSave) emptyBlocks.push_back(tokenOfStmt(raw.header));

      block.opCount = static_cast<int>(out.ops.size()) - block.opStart;
      out.blocks.push_back(std::move(block));
    }

    reportUnusedTemplates(parsed.templates, out.diagnostics);
    // A script with an error never runs, so "does nothing" would only add noise.
    if (!hasErrors(out.diagnostics))
      for (const Token& header : emptyBlocks)
        out.diagnostics.push_back(makeDiag(Severity::WARNING, "W_EMPTY_BLOCK", header,
                                           "this @source block does nothing"));
    return out;
  }

}  // namespace stencil::core::script
